src/Crypto.ts

```diff
@@ -1,150 +1,152 @@
-      buffer = Buffer.from(this.cipher.update(plainBuf))
-    }
-    return buffer
-  }
+    // Random salt
+    const salt = this.cryptoLib.randomBytes(cryptoConstants.SALT_LEN)
 
-  decryptSync(cipherBuf: Buffer, initVector?: Buffer): Buffer {
-    if (initVector && (initVector !== this.initVector)) {
-      this.resetDecipherSync(initVector!)
-    }
+    // Reset cipher
+    this.resetSync(this.masterKey, initVector, salt)
 
-    return Buffer.from(this.decipher.update(cipherBuf))
-  }
+    const encrypted = Buffer.concat([this.cipher.update(blockBuffer), this.cipher.final()])
 
-  async decrypt(cipherBuf: Buffer, initVector: Buffer | undefined = undefined): Promise<Buffer> {
-    if (initVector && (initVector !== this.initVector)) {
-      await this.resetDecipher(initVector!)
-    }
+    // Extract the auth tag
+    const authTag = this.cipher.getAuthTag()
 
-    let buffer: Buffer
-    if (this.useWebWorkers && this.cryptorWorker) {
-      buffer = await this.cryptorWorker.decryptBuf(cipherBuf)
-    } else {
-      buffer = Buffer.from(this.decipher.update(cipherBuf))
-    }
-    return buffer
+    // Construct chunk
+    return Buffer.concat([salt, initVector, authTag, encrypted])
   }
 
-  decryptCommitSync(): Buffer {
-    return Buffer.from(this.decipher.final())
-  }
+  async encryptBlock(blockBuffer: Buffer): Promise<Buffer> {
+    // Random initialization vector
+    const initVector = this.cryptoLib.randomBytes(cryptoConstants.INIT_VECTOR_LEN)
 
-  async decryptCommit(): Promise<Buffer> {
-    return Buffer.from(this.decipher.final())
-  }
+    // Random salt
+    const salt = this.cryptoLib.randomBytes(cryptoConstants.SALT_LEN)
 
-  // ========= HELPER FUNCTIONS =============
+    if (this.useWebWorkers) {
+      if (!(await this.waitForCryptoWorkerInit())) {
+        throw(Error('CryptoWorker does not exist'))
+      }
+      // Construct chunk
+      return Buffer.from(await this.cryptoWorker!.encryptBlock(blockBuffer, this.masterKey, this.algorithm, salt, initVector))
+    } else {
+      // Reset cipher
+      this.resetSync(this.masterKey, initVector, salt)
 
-  public getInitVector() : Buffer {
-    return this.initVector
-  }
+      // Encrypt blockBuffer
+      const encrypted = Buffer.concat([this.cipher.update(blockBuffer), this.cipher.final()])
 
-  private resetCipherSync(initVector: Buffer) {
-    this.initVector = initVector
-    this.cipher = crypto.createCipheriv(this.algorithm, this.key, initVector)
+      // Extract the auth tag
+      const authTag = this.cipher.getAuthTag()
 
-    return
-  }
-
-  private async resetCipher(initVector: Buffer) {
-    this.initVector = initVector
-    if (this.useWebWorkers && this.cryptorWorker) {
-      return await this.cryptorWorker.init(this.algorithm, this.key, initVector)
-    } else {
-      this.cipher = crypto.createCipheriv(this.algorithm, this.key, this.initVector)
+      // Construct chunk
+      return Buffer.concat([salt, initVector, authTag, encrypted])
```

Comment:
A "packet" diagram for this would be useful.

Reply:
I've added a chunk diagram to the beginning of the crypto class